burn-core/src/nn/rnn/lstm.rs

```diff
@@ -120,45 +134,45 @@ impl<B: Backend> Lstm<B> {
     pub fn forward(
         &self,
         batched_input: Tensor<B, 3>,
-        state: Option<(Tensor<B, 2>, Tensor<B, 2>)>,
+        state: Option<(Tensor<B, 3>, Tensor<B, 3>)>,
     ) -> (Tensor<B, 3>, Tensor<B, 3>) {
-        let [batch_size, seq_length, _] = batched_input.shape().dims;
-        let mut batched_cell_state = Tensor::zeros([batch_size, seq_length, self.d_hidden]);
-        let mut batched_hidden_state = Tensor::zeros([batch_size, seq_length, self.d_hidden]);
+        let num_directions = if let (Some(_), Some(_), Some(_), Some(_)) = (
+            &self.input_gate_bw,
+            &self.forget_gate_bw,
+            &self.output_gate_bw,
+            &self.cell_gate_bw,
+        ) {
+            2
+        } else {
+            1
+        };
+
+        let device = batched_input.clone().device();
 
-        let (mut cell_state, mut hidden_state) = match state {
-            Some((cell_state, hidden_state)) => (cell_state, hidden_state),
+        let [batch_size, seq_length, _] = batched_input.shape().dims;
+        let mut batched_cell_state =
+            Tensor::zeros_device([batch_size, seq_length, self.d_hidden * num_directions], &device);
```

Comment:
@nathanielsimard Is this necessary? The implementation is already tied to a device vis-à-vis the backend right?

Reply:
In my understanding, `Module::fork` can only fork the tensor in the module struct field to the given device. So, when creating a new Tensor in the `forward()`, it is necessary to specify the device?

By the way, now I feel that bidirectional LSTM may not be a common requirement for everyone. If someone needs bidirectional LSTM or multi-layer LSTM, it's probably best for them to implement it themselves. Additionally, keeping the implementation of Lstm unchanged is essential to avoid disrupting code compatibility.

Reply:
The device is necessary when creating a new tensor; otherwise, the new tensor will be created on the default device, but not necessarily on the same device as the module.